pkg/controller/serviceexternalip/controller.go

```diff
@@ -0,0 +1,394 @@
+		if apimachineryerrors.IsNotFound(err) {
+			c.deleteExternalIP(key)
+			return nil
+		}
+		return err
+	}
+
+	service := prev.DeepCopy()
+	// Service does not need external IP or type has changed.
+	if service.Spec.Type != corev1.ServiceTypeLoadBalancer {
+		c.deleteExternalIP(key)
+		return nil
+	}
+
+	currentIPPool := service.ObjectMeta.Annotations[antreaagenttypes.ServiceExternalIPPoolAnnotationKey]
+	prevIPAllocation, allocationExists := c.getAssignedExternalIPAllocation(key)
+	currentExternalIP := getServiceExternalIP(service)
+
+	// If user specifies external IP in spec, we should check whether it matches the current external IP.
+	specIPMatched := service.Spec.LoadBalancerIP == "" || service.Spec.LoadBalancerIP == currentExternalIP
+
+	if allocationExists && specIPMatched &&
+		c.externalIPAllocator.IPPoolExists(currentIPPool) &&
+		c.externalIPAllocator.IPPoolHasIP(currentIPPool, prevIPAllocation.ip) &&
+		currentIPPool == prevIPAllocation.ipPool &&
+		currentExternalIP == prevIPAllocation.ip.String() {
+		return nil
+	}
+
+	// The ExternalIPPool does not exist or has been deleted. Reclaim the external IP.
+	if currentIPPool != "" && !c.externalIPAllocator.IPPoolExists(currentIPPool) {
+		c.deleteExternalIP(key)
+		if currentExternalIP != "" {
+			service.Status.LoadBalancer.Ingress = nil
+			return c.updateService(prev, service)
+		}
+	}
+
+	// the external IP or ExternalIPPool changed somehow. Delete the previous allocation.
+	c.deleteExternalIP(key)
+	var newIPPool string
+	var newExternalIP net.IP
+	if service.Spec.LoadBalancerIP != "" {
+		// find the coressponding ExternalIPPool for user specified IP address.
+		newExternalIP = net.ParseIP(service.Spec.LoadBalancerIP)
+		newIPPool, err = c.externalIPAllocator.LocateIP(newExternalIP)
+		if err == nil {
+			err = c.externalIPAllocator.UpdateIPAllocation(newIPPool, newExternalIP)
+		}
+	} else if currentIPPool == "" {
+		// ExternalIPPool is not specified. Allocate IP and ExternalIPPool.
+		newIPPool, newExternalIP, err = c.externalIPAllocator.AllocateIP()
+	} else {
+		// Allocate IP from existing ExternalIPPool.
+		newExternalIP, err = c.externalIPAllocator.AllocateIPFromPool(currentIPPool)
+		newIPPool = currentIPPool
+	}
+	if err != nil {
+		// If the ExternalIPPool does not exist, we can ignore the error sine the Service will get requeued by ExternalIPPool change events.
+		if errors.Is(err, externalippool.ErrExternalIPPoolNotFound) {
+			klog.Errorf("Error when allocating IP from ExternalIPPool %s for Service %s: %v", newIPPool, key, err)
+			return nil
+		}
+		return fmt.Errorf("error when allocating IP %s from ExternalIPPool %s for Service %s: %v", newExternalIP, newIPPool, key, err)
+	}
+	klog.InfoS("Allocated external IP for service", "service", key, "ip", newExternalIP, "externalIPPool", newIPPool)
+	if service.Annotations == nil {
+		service.Annotations = make(map[string]string)
+	}
+	service.Annotations[antreaagenttypes.ServiceExternalIPPoolAnnotationKey] = newIPPool
```

Comment:
I thought the annotation will only be set by user to express their intention.
If user sets IP in spec and pool annotation together and they don't match, the code will override the pool? If yes, I think this may not be wanted. Not sure if we should really support allocating external IP without pool annotation. If we do this, we will need to provide a toggle to turn off the feature and keep it disabled by default when it graduates to beta, otherwise it may affect LB service while user want to use other LBs.
Any thoughts? @jianjuns

Reply:
I think if IP is not in the specified pool, we should just just fail. Also feel we should not annotate IPPool to a Service. The annotations should be managed by users.

Reply:
Currently, the annotation will be overwritten for the following scenarios:
1. The annotation for IP pool is empty.
2. User specifies an IP in Service spec, but it does not match the existing annotation.

For the other cases, the controller will allocate from the given IP pool and fail if allocation fails. Do we want to keep the feature of not specifying any IP pool and allocate one for users?

Reply:
I removed the function of updating annotations here. So service with no IP pool annotations or mismatching annotation and specified IP in the spec will not be assigned an external IP.